velox/functions/prestosql/DecimalArithmetic.cpp

```diff
@@ -0,0 +1,249 @@
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row], rawValues[row], constant, aRescale_, bRescale_);
+      });
+    } else if (args[0]->isFlatEncoding() && args[1]->isFlatEncoding()) {
+      auto flatA = args[0]->asUnchecked<FlatVector<A>>();
+      auto rawA = flatA->mutableRawValues();
+      auto flatB = args[1]->asUnchecked<FlatVector<B>>();
+      auto rawB = flatB->mutableRawValues();
+
+      R* rawResults = nullptr;
+      if (!(*result)) {
+        // Try to reuse one of the flat vectors to store result.
+        if (resultType->kind() == flatA->typeKind() &&
+            BaseVector::isReusableFlatVector(args[0])) {
+          rawResults = reinterpret_cast<R*>(rawA);
+          *result = std::move(args[0]);
+        } else if (
+            resultType->kind() == flatB->typeKind() &&
+            BaseVector::isReusableFlatVector(args[1])) {
+          rawResults = reinterpret_cast<R*>(rawB);
+          *result = std::move(args[1]);
+        }
+      }
+      if (!rawResults) {
+        rawResults = prepareResults(rows, resultType, context, result);
+      }
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row], rawA[row], rawB[row], aRescale_, bRescale_);
+      });
+    } else {
+      exec::DecodedArgs decodedArgs(rows, args, context);
+      auto a = decodedArgs.at(0);
+      auto b = decodedArgs.at(1);
+      BaseVector::ensureWritable(rows, resultType, context->pool(), result);
+      auto rawResults =
+          (*result)->asUnchecked<FlatVector<R>>()->mutableRawValues();
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row],
+            a->valueAt<A>(row),
+            b->valueAt<B>(row),
+            aRescale_,
+            bRescale_);
+      });
+    }
+  }
+
+ private:
+  R* prepareResults(
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const {
+    BaseVector::ensureWritable(rows, resultType, context->pool(), result);
+    (*result)->clearNulls(rows);
+    return (*result)->asUnchecked<FlatVector<R>>()->mutableRawValues();
+  }
+
+  R* getRawResults(
+      const SelectivityVector& rows,
+      VectorPtr& flat,
+      const TypePtr& resultType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const {
+    // Check if input can be reused for results.
+    R* rawResults;
+    if (!(*result) && flat->typeKind() == resultType->kind() &&
+        BaseVector::isReusableFlatVector(flat)) {
+      auto flatValues = flat->asUnchecked<FlatVector<R>>();
+      rawResults = flatValues->mutableRawValues();
+      *result = std::move(flat);
+    } else {
+      rawResults = prepareResults(rows, resultType, context, result);
+    }
+    return rawResults;
+  }
+  uint8_t aRescale_;
```

Comment:
Can these be const?

nit: add empty line before member variables.

Reply:
Done.